### DISCUSSION TITLE
Running test stage after training stage using LightningCLI

### DISCUSSION THREAD

**Discussion by SinaSarparast** [Lightning-AI/pytorch-lightning#15541] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hello,

I'm using LightningCLI for training a model. I want to run training, validation and testing all in the same experiment to have them all logged in one place, so that they are easily accessible by the logger (mlflow). The validation stage happens after each epoch and the parameters are logged. The testing stage does not take place after training automatically.

I know that I can use fit,validate,test,predict,tune as the documentation says.

`$ python trainer.py --help
usage: trainer.py [-h] [--config CONFIG] [--print_config [={comments,skip_null}+]] {fit,validate,test,predict,tune} ...
`

After the training is over the execution of the experiment is terminated and I have manually execute the code and specify the path to the checkpoint. Therefore, a new experiment is created and the test metrics are logged independently as a separate experiment. I want to run the test stage right after training is over with the current model and avoid running a new experiment for testing. I couldn't find a way to do this with the Lightning CLI. How can I run the test stage right after training is over?

Thanks,

**Comment by tshu-w** [comment, marked as answer]
You can [Customize the LightningCLI](https://pytorch-lightning.readthedocs.io/en/latest/cli/lightning_cli_expert.html#customize-the-lightningcli) `after_fit`, here is an [example](https://github.com/tshu-w/lightning-template/blob/cacec51c9035d614e9a92a158bc1ef683e7b5b3e/src/utils/lit_cli.py#L54-L100).
Or implement a [callback](https://github.com/tshu-w/lightning-template/blob/main/src/callbacks/metric.py) and add it to LightningCLI by default.
